cmd/deploy/input_test.go

```diff
@@ -2,9 +2,240 @@ package main
+	"github.com/docker-library/meta-scripts/registry"
 )
 
+func TestGetRefs(t *testing.T) {
+	type want struct {
+		digest ociregistry.Digest
+		refs   []registry.Reference
+	}
+	type fixture struct {
+		name       string
+		rawDigests []string
+		deployType deployType
+		want       want
+		wantErr    string
+	}
+	for _, x := range []fixture{
+		{
+			"happy path",
+			[]string{
+				"localhost:5000/foo@sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+				"localhost:5000/bar:some-tag",
+				"localhost:5000/baz",
+			},
+			typeManifest,
+			want{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+				[]registry.Reference{
+					{
+						Host:       "localhost:5000",
+						Repository: "foo",
+						Digest:     "sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+					},
+					{
+						Host:       "localhost:5000",
+						Repository: "bar",
+						Tag:        "some-tag",
+					},
+					{
+						Host:       "localhost:5000",
+						Repository: "baz",
+					},
+				},
+			},
+			"",
+		},
+		{
+			"invalid digest",
+			[]string{
+				"localhost:5000/foo@bad_digest",
+			},
+			typeManifest,
+			want{},
+			"localhost:5000/foo@bad_digest: failed to parse ref: invalid digest \"bad_digest\": invalid checksum digest format",
+		},
+		{
+			"invalid host",
+			[]string{
+				"...",
+			},
+			typeManifest,
+			want{},
+			"...: failed to parse ref: invalid reference syntax (\"...\")",
+		},
+		{
+			"invalid tag",
+			[]string{
+				"localhost:5000/foo:#",
+			},
+			typeManifest,
+			want{},
+			"localhost:5000/foo:#: failed to parse ref: tag \"#\" does not start with word character",
+		},
+		{
+			"invalid reference url",
+			[]string{
+				"localhost:5000/foo?test",
+			},
+			typeManifest,
+			want{},
+			"localhost:5000/foo?test: failed to parse ref: tag \"5000/foo?test\" contains invalid invalid character '/'",
+		},
+		{
+			"mismatch digest",
+			[]string{
+				"localhost:5000/foo@sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+				"localhost:5000/foo@sha256:00042f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+			},
+			typeManifest,
+			want{},
+			"refs digest mismatch in localhost:5000/foo@sha256:00042f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d: sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+		},
+		{
+			"can't push to blob",
+			[]string{
+				"localhost:5000/foo:tag",
+			},
+			typeBlob,
+			want{},
+			"cannot push blobs to a tag: localhost:5000/foo:tag",
+		},
+	} {
+		t.Run(x.name, func(t *testing.T) {
+			refsDigest, refs, err := getRefs(x.rawDigests, x.deployType)
+
+			if x.wantErr != "" && (err == nil || err.Error() == "") {
+				t.Fatalf("Expected error not returned: %s", x.wantErr)
+			}
+
+			if x.wantErr != "" && err.Error() != x.wantErr {
+				t.Fatalf("Expected error doesn't match. got:\n'%s',\n\nexpected:\n'%s'", err.Error(), x.wantErr)
+			}
```

Comment:
Checking whether we got an error object but that it returns the empty string for the error message is a bit strange -- I also think this code would be easier to follow if it had `if x.wantErr != ""` only once, and a sub-`if` for `if err == nil { fatal(unexpected errorless return) }` (which should maybe also include the other things we _did_ get in the error so it's easier to narrow down where it's failing) + something like `if errStr := err.Error(); !strings.Contains(errStr, x.wantErr) { fatal(unexpected err (looking for %q): errStr, x.wantErr) }`